controllers/machine_controller.go

```diff
@@ -242,6 +255,73 @@ func (r *MachineReconciler) isDeleteNodeAllowed(ctx context.Context, machine *cl
 	}
 }
 
+func (r *MachineReconciler) drainNode(ctx context.Context, cluster *clusterv1.Cluster, name string) error {
+	var kubeClient kubernetes.Interface
+	if cluster == nil {
+		var err error
+		kubeClient, err = kubernetes.NewForConfig(r.config)
+		if err != nil {
+			return fmt.Errorf("unable to build kube client: %v", err)
+		}
+	} else {
+		// Otherwise, proceed to get the remote cluster client and get the Node.
+		remoteClient, err := remote.NewClusterClient(r.Client, cluster)
+		if err != nil {
+			klog.Errorf("Error creating a remote client for cluster %q while deleting Machine %q, won't retry: %v",
+				cluster.Name, name, err)
+			return nil
+		}
+		var err2 error
+		kubeClient, err2 = kubernetes.NewForConfig(remoteClient.RESTConfig())
+		if err2 != nil {
+			klog.Errorf("Error creating a remote client for cluster %q while deleting Machine %q, won't retry: %v",
+				cluster.Name, name, err)
+			return nil
+		}
+	}
+
+	node, err := kubeClient.CoreV1().Nodes().Get(name, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// If an admin deletes the node directly, we'll end up here.
+			klog.Infof("Could not find node from noderef, it may have already been deleted: %v", cluster.Name)
+			return nil
+		}
+		return fmt.Errorf("unable to get node %q: %v", name, err)
+	}
+
+	drainer := &kubedrain.Helper{
+		Client:              kubeClient,
+		Force:               true,
+		IgnoreAllDaemonSets: true,
+		DeleteLocalData:     true,
+		GracePeriodSeconds:  -1,
+		// If a pod is not evicted in 20 second, retry the eviction next time the
+		// machine gets reconciled again (to allow other machines to be reconciled)
+		Timeout:               20 * time.Second,
+		OnPodDeletedOrEvicted: onPodDeletedOrEvicted,
+		Out:                   Writer{klog.Info},
+		ErrOut:                Writer{klog.Error},
+		DryRun:                false,
+	}
+
+	if err := kubedrain.RunCordonOrUncordon(drainer, node, true); err != nil {
+		// Machine still tries to terminate after drain failure
+		klog.Errorf("%q: node cordon failed %q; %v", cluster.Name, name, err)
+		return fmt.Errorf("unable to cordon node %s: %v", node.Name, err)
```

Comment:
```suggestion
		return errors.Errorf("unable to cordon node %s: %v", node.Name, err)
```

Reply:
Do we want to record an event here against the Machine so the user is informed?

Reply:
Well, I think in normal operation, this would just be due to temporary api/network type issue.  It should not happen during normal operations, if you have permissions to delete a node, I think you have permissions to cordon.